src/passes/Outlining.cpp

```diff
@@ -0,0 +1,211 @@
+        auto& child = block->list[blockIterator.index];
+        [[maybe_unused]] auto name = getExpressionName(child);
+        // std::cout << "Pushing a task to call StringifyWalker::scan with block
+        // list item: " << name  << " " << child << std::endl;
+        stringify->pushTask(StringifyWalker::scan, &child);
+      }
+      break;
+    }
+    case Expression::Id::IfId: {
+      auto* iff = curr->dynCast<If>();
+      stringify->pushTask(StringifyWalker::scan, &iff->ifFalse);
+      // std::cout << "Pushing an task to call StingifyWalker::scan on ifFalse "
+      // << std::endl;
+      stringify->pushTask(StringifyWalker::scan, &iff->ifTrue);
+      // std::cout << "Pushing an task to call StingifyWalker::scan on ifTrue "
+      // << std::endl;
+      stringify->pushTask(StringifyWalker::scan, &iff->condition);
+      // std::cout << "Pushing an task to call StingifyWalker::scan on
+      // ifCondition " << std::endl;
+      break;
+    }
+    case Expression::Id::TryId: {
+      auto* tryy = curr->dynCast<Try>();
+      auto blockIterator = tryy->catchBodies.end();
+      while (blockIterator != tryy->catchBodies.begin()) {
+        blockIterator--;
+        auto& child = tryy->catchBodies[blockIterator.index];
+        [[maybe_unused]] auto name = getExpressionName(child);
+        // std::cout << "Pushing a task to call StringifyWalker::scan with try
+        // catchBody: " << name  << " " << child << std::endl;
+        stringify->pushTask(StringifyWalker::scan, &child);
+      }
+      stringify->pushTask(StringifyWalker::scan, &tryy->body);
+      break;
+    }
+    case Expression::Id::LoopId: {
+      auto* loop = curr->dynCast<Loop>();
+      stringify->pushTask(StringifyWalker::scan, &loop->body);
+      break;
+    }
+    default: {
+      auto name = getExpressionName(*currp);
+      std::cout << "scanChildren reached an unimplemented expression: " << name
+                << std::endl;
+    }
+  }
+}
+
+void StringifyWalker::handler(StringifyWalker* stringify, Expression**) {
+  // printf("In StringifyWalker::handler\n");
+  auto& queue = stringify->queue;
+  if (!queue.empty()) {
+    stringify->pushTask(StringifyWalker::handler, nullptr);
+    Expression** currp = queue.front();
+    queue.pop();
+    [[maybe_unused]] auto name = getExpressionName(*currp);
+    // std::cout << "queue has an item, " << name << std::endl;
+    StringifyWalker::scanChildren(stringify, currp);
+  } else {
+    // std::cout << "queue is empty" << std::endl;
+  }
+}
+
+void StringifyWalker::scan(StringifyWalker* self, Expression** currp) {
+  Expression* curr = *currp;
+  [[maybe_unused]] auto name = getExpressionName(curr);
+  // std::cout << "StringifyWalker::scan() on: " << name << std::endl;
+  // curr->dump();
+
+  switch (curr->_id) {
+    case Expression::Id::BlockId:
+    case Expression::Id::IfId:
+    case Expression::Id::LoopId:
+    case Expression::Id::TryId: {
+      self->visitControlFlow(self, currp);
```

Comment:
I wonder if this could just be a `visitExpression(curr)` call. Then we would leave it up to the client code implementing `visitExpression` to figure out what to do differently with control flow structures, if anything.